waku/v2/peermanager/topic_event_handler_test.go

```diff
@@ -0,0 +1,218 @@
+	require.NoError(t, err)
+
+	// Wait for event loop to call handler
+	time.Sleep(200 * time.Millisecond)
+
+	// Check the original topic was removed from Peer Manager
+	pm.topicMutex.RLock()
+	_, ok = pm.subRelayTopics[pubSubTopic]
+	require.False(t, ok)
+	pm.topicMutex.RUnlock()
+
+	r.Stop()
+}
+
+func TestHandlePeerTopicEvent(t *testing.T) {
+	log := utils.Logger()
+	pubSubTopic := "/waku/2/go/pm/test"
+	ctx := context.Background()
+
+	hosts := make([]host.Host, 5)
+	relays := make([]*relay.WakuRelay, 5)
+
+	for i := 0; i < 5; i++ {
+		relays[i], hosts[i], _ = makeWakuRelay(t, log)
+		err := relays[i].Start(ctx)
+		require.NoError(t, err)
+	}
+
+	// Create peer manager instance with the first hosts
+	pm, _ := makePeerManagerWithEventBus(t, relays[0], &hosts[0])
+	pm.ctx = ctx
+	pm.RegisterWakuProtocol(relay.WakuRelayID_v200, relay.WakuRelayENRField)
+
+	// Connect host[0] with all other hosts to reach 4 connections
+	for i := 1; i < 5; i++ {
+		pm.host.Peerstore().AddAddrs(hosts[i].ID(), hosts[i].Addrs(), peerstore.PermanentAddrTTL)
+		err := pm.host.Connect(ctx, hosts[i].Peerstore().PeerInfo(hosts[i].ID()))
+		require.NoError(t, err)
+		err = pm.host.Peerstore().(wps.WakuPeerstore).SetDirection(hosts[i].ID(), network.DirOutbound)
+		require.NoError(t, err)
+
+	}
+
+	// Wait for connections to settle
+	time.Sleep(2 * time.Second)
+
+	if len(pm.host.Peerstore().(*wps.WakuPeerstoreImpl).PeersByPubSubTopic(pubSubTopic)) == 0 {
+		log.Info("No peers for the topic yet")
+	}
+
+	for i := 1; i < 5; i++ {
+		// Subscribe to Pubsub topic on first host only
+		_, err := relays[i].Subscribe(ctx, protocol.NewContentFilter(pubSubTopic))
+		require.NoError(t, err)
+	}
+
+	// Start event loop to listen to events
+	ctxEventLoop := context.Background()
+	go pm.peerEventLoop(ctxEventLoop)
+
+	//// Prepare emitter
+	//emitter, err := eventBus.Emitter(new(relay.EvtPeerTopic))
+	//require.NoError(t, err)
+
+	//// Send PEER_JOINED events for hosts 2-5
+	//for i := 1; i < 5; i++ {
+	//	err = emitTopicEvent(pubSubTopic, hosts[i].ID(), emitter, relay.PEER_JOINED)
+	//	require.NoError(t, err)
+	//	time.Sleep(100 * time.Millisecond)
+	//}
+
+	// Wait for connections to settle
+	time.Sleep(30 * time.Second)
+
+	// Check four hosts have joined the topic
+	require.Equal(t, 4, len(pm.host.Peerstore().(*wps.WakuPeerstoreImpl).PeersByPubSubTopic(pubSubTopic)))
+
+	// Check all hosts have been connected
+	for _, peer := range pm.host.Peerstore().(*wps.WakuPeerstoreImpl).PeersByPubSubTopic(pubSubTopic) {
+		require.Equal(t, network.Connected, pm.host.Network().Connectedness(peer))
+	}
+
+	//// Send PEER_LEFT events for hosts 2-5
+	//for i := 1; i < 5; i++ {
+	//	err = emitTopicEvent(pubSubTopic, hosts[i].ID(), emitter, relay.PEER_LEFT)
+	//	require.NoError(t, err)
+	//	time.Sleep(100 * time.Millisecond)
+	//}
+	//
+	//// Check all hosts have left the topic
+	//require.Equal(t, 0, len(pm.host.Peerstore().(*wps.WakuPeerstoreImpl).PeersByPubSubTopic(pubSubTopic)))
```

Comment:
This can also be uncommented once you call unsubscribe

Reply:
Let's see later.